lib/ansible/executor/task_executor.py

```diff
@@ -1051,22 +1051,23 @@ def _set_connection_options(self, variables, templar):
 
         task_keys = self._task.dump_attrs()
 
-        # The task_keys 'timeout' attr is the task's timeout, not the connection timeout.
-        # The connection timeout is threaded through the play_context for now.
-        task_keys['timeout'] = self._play_context.timeout
-
         if self._play_context.password:
             # The connection password is threaded through the play_context for
             # now. This is something we ultimately want to avoid, but the first
             # step is to get connection plugins pulling the password through the
             # config system instead of directly accessing play_context.
             task_keys['password'] = self._play_context.password
 
-        # Prevent task retries from overriding connection retries
+        # Prevent task retries/timeout from overriding connection retries
         del(task_keys['retries'])
+        del(task_keys['timeout'])
 
         # set options with 'templated vars' specific to this plugin and dependent ones
         self._connection.set_options(task_keys=task_keys, var_options=options)
+        if 'timeout' in self._connection._options and self._connection.get_option('timeout') is None:
```

Comment:
dont inspect _options, get the plugin config defs and see if 'timeout' is one of the options (also allows you to check for cli/env/ini entries)

Reply:
I wonder if it should just be a warning/error for a plugin to have anything like:
```
cli:
  - timeout  # defaults to DEFAULT_TIMEOUT, which defaults to 10
default: anything that isn't 10
```
This PR is special-casing --timeout, but it would be the same for any cli option that has a default value.

Also wondering what should happen if a plugin has multiple items in an option's ``cli`` list? If one of those has a default value, it won't ever consider the rest in the list currently. Maybe that should be a warning/error too.

Reply:
i would move from using the default in the cli options and allowing the plugins to diverge, eventually removing the 'global options'